src/github.com/getlantern/balancer/dialer.go

```diff
@@ -94,44 +78,94 @@ func (d *dialer) start() {
 					d.OnClose()
 				}
 				return
-			case t := <-d.errCh:
-				lastFailed = t
-			case <-timer.C:
-				ok := d.Check()
-				if ok {
-					lastCheckSucceeded = time.Now()
-					timer.Reset(longDuration)
-					consecCheckFailures = 0
-				} else {
-					consecCheckFailures += 1
-				}
+			case <-d.errCh:
+				d.markFailure()
+			case <-d.checkTimer.C:
+				go func() {
```

Comment:
If checking takes longer than maxCheckTimeout, we will start piling up goroutines/connections.  Previously, checks happened on only one goroutine.  I think we should continue doing checks on just one goroutine.